src/core/qgsgrouplayer.cpp

```diff
@@ -0,0 +1,379 @@
+/***************************************************************************
+    qgsgrouplayer.cpp
+    ----------------
+  Date                 : September 2021
+  Copyright            : (C) 2021 by Nyall Dawson
+  Email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsgrouplayer.h"
+#include "qgsmaplayerfactory.h"
+#include "qgspainting.h"
+#include "qgsmaplayerlistutils.h"
+#include "qgsgrouplayerrenderer.h"
+#include "qgsmaplayerref.h"
+#include "qgsvectorlayer.h"
+#include "qgscoordinatetransform.h"
+#include "qgspainteffect.h"
+#include "qgsmessagelog.h"
+#include "qgspainteffectregistry.h"
+#include "qgsapplication.h"
+
+QgsGroupLayer::QgsGroupLayer( const QString &name, const LayerOptions &options )
+  : QgsMapLayer( QgsMapLayerType::GroupLayer, name )
+  , mTransformContext( options.transformContext )
+{
+  mShouldValidateCrs = false;
+  mValid = true;
+
+  QgsDataProvider::ProviderOptions providerOptions;
+  providerOptions.transformContext = options.transformContext;
+  mDataProvider = new QgsGroupLayerDataProvider( providerOptions, QgsDataProvider::ReadFlags() );
+}
+
+QgsGroupLayer::~QgsGroupLayer()
+{
+  emit willBeDeleted();
+  delete mDataProvider;
+}
+
+QgsGroupLayer *QgsGroupLayer::clone() const
+{
+  const QgsGroupLayer::LayerOptions options( mTransformContext );
+  std::unique_ptr< QgsGroupLayer > layer = std::make_unique< QgsGroupLayer >( name(), options );
+  QgsMapLayer::clone( layer.get() );
+  layer->setChildLayers( _qgis_listRefToRaw( mChildren ) );
+  layer->setPaintEffect( mPaintEffect ? mPaintEffect->clone() : nullptr );
+  return layer.release();
+}
+
+QgsMapLayerRenderer *QgsGroupLayer::createMapRenderer( QgsRenderContext &context )
+{
+  return new QgsGroupLayerRenderer( this, context );
+}
+
+QgsRectangle QgsGroupLayer::extent() const
```

Comment:
@nyalldawson , not sure if that'd be worth it, but one could cache the extent here and invalidate whenever the layers contained without the group change.

Reply:
Let's see if it becomes an issue -- we might end up adding whole lot of complexity for no noticeable benefit

Reply:
sorry, coming late to the party, but there seems to be  `static QgsRectangle combinedExtent( const QList<QgsMapLayer *> &layers, const QgsCoordinateReferenceSystem &crs, const QgsCoordinateTransformContext &transformContext )` in `QgsMapLayerUtils`.
https://github.com/qgis/QGIS/blob/master/src/core/qgsmaplayerutils.h

Could you use it or move this implementation to `QgsMapLayerUitls`?

Reply:
@3nids 
> Could you use it or move this implementation to QgsMapLayerUitls?

done in https://github.com/qgis/QGIS/pull/46183